sdk-rs/src/lib.rs

```diff
@@ -0,0 +1,790 @@
+//! Drift SDK
+
+use std::{collections::HashMap, sync::Arc};
+
+use anchor_lang::{AccountDeserialize, InstructionData, ToAccountMetas};
+use constants::{state_account, PerpMarketConfig, SpotMarketConfig};
+use drift_program::{
+    controller::position::PositionDirection,
+    math::constants::QUOTE_SPOT_MARKET_INDEX,
+    state::{
+        order_params::{ModifyOrderParams, OrderParams},
+        user::{MarketType, Order, OrderStatus, PerpPosition, SpotPosition, User},
+    },
+};
+use futures_util::stream::StreamExt;
+use solana_account_decoder::UiAccountEncoding;
+use solana_client::{
+    nonblocking::{pubsub_client::PubsubClient, rpc_client::RpcClient},
+    rpc_config::RpcAccountInfoConfig,
+};
+pub use solana_sdk::pubkey::Pubkey;
+use solana_sdk::{
+    account::{Account, AccountSharedData, ReadableAccount},
+    commitment_config::{CommitmentConfig, CommitmentLevel},
+    instruction::{AccountMeta, Instruction},
+    signature::{keypair_from_seed, Keypair, Signature},
+    signer::Signer,
+    transaction::Transaction,
+};
+use tokio::sync::{
+    watch::{self, Receiver},
+    RwLock,
+};
+
+pub mod constants;
+pub mod dlob;
+pub mod types;
+use types::*;
+pub mod utils;
+
+/// Drift Client API
+///
+/// Cheaply clonable
+#[derive(Clone)]
+pub struct DriftClient {
+    backend: &'static DriftClientBackend,
+}
+
+impl DriftClient {
+    pub async fn new(endpoint: &str) -> Result<Self, SdkError> {
+        Ok(Self {
+            backend: Box::leak(Box::new(DriftClientBackend::new(endpoint).await?)),
+        })
+    }
+    /// Transparently subscribe to account updates for a given account/sub-account, enabling more efficient, cached queries.
+    ///
+    /// `account` the drift user PDA
+    ///
+    /// This does not return anything but allows subsequent queries to benefit, useful for long-lived workloads expecting to query the same account frequently.
+    /// In contrast, the default behaviour is to _always_ fetch the account data via network request which maybe better for ad-hoc workloads.
+    pub async fn subscribe_account(&self, account: &Pubkey) -> Result<(), SdkError> {
```

Comment:
yeah it might be a good idea to traitify `DriftClientBackend` so the user can have the option between polling, websocket, grpc, etc... (once upon a time the public Solana RPC disabled websockets)

`account` here is also kind of ambiguous, maybe specify `User` since the backend is decoding that.

SDK users will also be interested in:
* market accounts (and corresponding serum + phoenix accounts for spot)
* oracle accounts
* perhaps multiple accounts
    * if using multiple subaccounts
    * maintaining a User/UserStats map (way more efficient to use getProgramAccounts/programSubscribe for these use cases)

Reply:
thanks for the feedback, I've put the account retrieval logic behind a trait